### ISSUE TITLE
Unable to use `$\a` to create the bell ASCII codepoint

### ISSUE THREAD

**Issue by gearnode** [erlang/otp#4844]
**Describe the bug**
It is not possible to use the `$\a` to create the `bell` ASCII codepoint.

**To Reproduce**
```erlang
$\a =:= 97.
```

**Expected behavior**
```erlang
$\a =:= 7.
```

**Affected versions**
At least OTP 23 and OTP 24.

**Comment by gearnode** [comment]
Another thing (which, from my point of view, is considered as a bug) is that not recognized codepoint are silently transformed to another codepoint. 

**To Reproduce**
```erlang
$\z =:= $z
```
**Expected behavior**
Raise a syntax error when codepoint is not valid.

**Comment by KennethL** [comment]
`$\a` is not documented as one of the allowed `$\x` combinations and then it fallbacks to the same value as $a which is 97.
The `BEL` ascii character has value the 7 and can be expressed as ctrl-g which in Erlang is represented as the integer 7 or in the $ notation as `$\^g`.

I don't understand what connection there is between the character `a` and the ascii (or unicode) code point 7.

**Comment by KennethL** [comment]
You can also create the BEL with `$\x07`. Or in a string like this `"\x07"` or like this `"\^g"`.

**Comment by frej** [comment]
> I don't understand what connection there is between the character a and the ascii (or unicode) code point 7.

In C89, `\a` stands for "alert" and is defined as `0x07`.

**Comment by galdor** [comment]
In other versions of the C standard too. And also in Java, Perl, Ruby, Python, Haskell… Since Erlang supports other common used alphabetic escape sequences, it would make sense to support this one to be coherent.

As for the fact that unknown alphabetic escape sequences fallback to the character without the escaping, I find it confusing: if the sequence is invalid, an error should be signaled. Silently ignoring invalid data is rarely a good thing.

**Comment by KennethL** [comment]
Yes but in this case this is request to add the `$\a` as a way to express the BEL (07) ascii code point.
And the behavior where unknown letters are used in the ? position here `$\?` are not reported as errors might have its background in that it has always been like that and would be incompatible if changed. Will check that.

**Comment by gearnode** [comment]
Erlang already supports `$\b`, `$\t`, `$\n`, `$\v`, `$\f` and `$\r`, it makes sense to support `$\a` too. And as @frej and @galdor says, this kind of escape sequence is supported in many other languages.

**Comment by garazdawi** [comment]
> Erlang already supports $\b, $\t, $\n, $\v, $\f and $\r, it makes sense to support $\a too. And as @frej and @galdor says, this kind of escape sequence is supported in many other languages.

If it had not broken backward compatibility I would agree, but since today `$\a` today gives you 97 we need to first consider if there is any chance that we break anyone's code.

If we do this change, we should also consider disallowing any of the `$\N` that are not defined in http://erlang.org/doc/reference_manual/data_types.html#escape-sequences. Unless there is a reason why things are the way they are that I do not realize.

**Comment by RaimoNiskanen** [comment]
Does this kind of change merit a compiler flag, warnings, and possibly changed default in next major release?
Still kind of dangerous to change the default behaviour to return 7 for $\a.  Silent semantics change; actually worse than silently accepting invalid data...

**Comment by galdor** [comment]
Maybe a compiler flag for both fixing `$\a` and signaling an error when the escape sequence is invalid ? After one or two releases it could become default.

**Comment by lhoguin** [comment]
I'd say just add a warning about unknown escape sequences, always enabled, and later on make it an error / add `$\a` and any other missing. It's very unlikely that users have purposefully used those.

**Comment by KennethL** [comment]
We discussed this in the OTP technical board and decided:
- Generate a warning for unsupported (= undocumented) escape characters like `$\a` in OTP 25.
- Generate an error for unsupported (= undocumented) escape characters (but not `$\a`) in OTP 26.
- Introduce `$\a` -> 7 (and possibly some other characters if they are common in other languages) in OTP 26
    
Today `$\a` and other unsupported escape characters are silently generating the same integer value as `$a` which we regard as a bug and definitely undefined behavior. But because of the risk that there are programs out there using it even if it is very unlikey, we will do this over 2 major releases with first warning and then error.

**Comment by gearnode** [comment]
Perfect!

Thanks for your reactivity!
